polaris/hub/storage.py

```diff
@@ -0,0 +1,431 @@
+
+            for page in page_iterator:
+                # Contents are "files"
+                for obj in page.get("Contents", []):
+                    key = obj["Key"][len(prefix) :]
+                    if key:
+                        yield key.split("/")[0]
+
+                # CommonPrefixes are "subdirectories"
+                for common_prefix in page.get("CommonPrefixes", []):
+                    yield common_prefix["Prefix"][len(prefix) :].strip("/")
+
+    def getitems(self, keys: Sequence[str], *, contexts: Mapping[str, Context]) -> dict[str, Any]:
+        """
+        More efficient implementation of getitems using concurrent fetching through multiple threads.
+
+        The default implementation uses __contains__ to check existence before fetching the value,
+        which doubles the number of requests.
+        """
+
+        def fetch_key(key: str) -> tuple[str, Any]:
+            """
+            Wrapper that looks up a key's value in the store
+            """
+            try:
+                return key, self[key]
+            except KeyError:
+                return key, None
+
+        results = {}
+        with ThreadPoolExecutor() as executor:
+            # Create a future for each key to look up
+            future_to_key = {executor.submit(fetch_key, key): key for key in keys}
+
+            # As each future completes, collect the results, if any
+            for future in as_completed(future_to_key):
+                key, value = future.result()
+                if value is not None:
+                    results[key] = value
+
+        return results
+
+    def getsize(self, key: str) -> int:
+        """
+        Return the size (in bytes) of the object at the given key.
+        """
+        with handle_s3_errors():
+            response = self.s3_client.head_object(Bucket=self.bucket_name, Key=self._full_key(key))
+            return response["ContentLength"]
+
+    ## MutableMapping implementation, expected by Zarr
+
+    def __getitem__(self, key: str) -> bytes:
+        """
+        Retrieves the value for the given key from the store.
+
+        Makes no provision to handle overly large values returned.
+        """
+        with handle_s3_errors():
+            try:
+                full_key = self._full_key(key)
+                response = self.s3_client.get_object(Bucket=self.bucket_name, Key=full_key)
+                return response["Body"].read()
+            except self.s3_client.exceptions.NoSuchKey:
+                raise KeyError(key)
+
+    def __setitem__(self, key: str, value: bytes | bytearray | memoryview) -> None:
+        """
+        Persists the given value in the store.
+
+        Based on value size, will use multipart upload for large files,
+        or a single put_object call.
+        """
+        if isinstance(value, memoryview):
+            value = value.tobytes()
+
+        if len(value) > self.part_size:
```

Comment:
Yeah, that's a good question.

I thought about it when writing these methods, and I wanted `_multipart_upload` to be coherent by itself and low-coupling. If I have the caller pass in the hash, then I've coupled that caller's code with `_multipart_upload`'s internal need to use MD5. If we switch over to another hash algorithm, or some other integrity mechanism, then the caller needs to be updated as well.

I weighted that against the code duplication. I figured since it's a call to a standard library, then that was acceptable as a tradeoff for lower coupling. YMMV.

John Ousterhout has a great little book about this kind of things. 😄